webpushd should protect incoming pushes with an os_transaction

bnham opened the pull request:
#### f7b5d128dffff0d481dbb976e6c156f79b73e180
<pre>
webpushd should protect incoming pushes with an os_transaction
<a href="https://bugs.webkit.org/show_bug.cgi?id=241740">https://bugs.webkit.org/show_bug.cgi?id=241740</a>

Reviewed by Youenn Fablet.

webpushd needs to acquire an os_transaction in its incoming push message path, or else it might get
jetsamed in the middle of processing a push.

* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::didReceivePushMessage): Create a transaction around fetching the decryption
key from the DB and decrypting the push.

* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::Daemon):
(WebPushD::Daemon::startIncomingPushTransactionTimer):
(WebPushD::Daemon::cancelIncomingPushTransactionTimer):
(WebPushD::Daemon::incomingPushTransactionTimerFired):
(WebPushD::Daemon::handleIncomingPush):
(WebPushD::Daemon::getPendingPushMessages): Create a transaction around waking up the UIProcess to
handle the decrypted push and time it out after 10 seconds.

* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain): Create a transaction around the time from when we spawn to when we
start initializing PushService. (PushService initialization involves async steps and takes care of
acquiring its own transaction in its constructor.)

Canonical link: <a href="https://commits.webkit.org/251879@main">https://commits.webkit.org/251879@main</a>
</pre>

youennf commented on Source/WebKit/webpushd/PushService.mm:
```diff
@@ -691,6 +693,8 @@ void reject(WebCore::ExceptionData&& data)
 
         completionHandler();
+
+        transaction = nullptr;
```
Why do we need transaction = nullptr here while we do not need it in the return early case?

bnham replied:
The RetainPtr going out of scope should be sufficient to release the `os_transaction` in both cases. I put the store to nullptr here just to suppress the unused variable warning. Maybe `UNUSED_PARAM` would be better?

youennf commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -607,6 +633,8 @@ void handleWebPushDMessage(ClientConnection* connection, Span<const uint8_t> enc
 
 void Daemon::handleIncomingPush(const String& bundleIdentifier, WebKit::WebPushMessage&& message)
 {
+    startIncomingPushTransactionTimer();
```
I would go with ensureIncomingPushTransaction(). The timer is really a detail.

youennf commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -666,6 +696,9 @@ void handleWebPushDMessage(ClientConnection* connection, Span<const uint8_t> enc
     replySender(WTFMove(resultMessages));
+    
+    if (m_pushMessages.isEmpty())
+        cancelIncomingPushTransactionTimer();
```
I would go with releaseIncomingPushTransaction() as a name.

youennf commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -330,6 +337,25 @@ void handleWebPushDMessage(ClientConnection* connection, Span<const uint8_t> enc
+void Daemon::incomingPushTransactionTimerFired()
+{
+    RELEASE_LOG_ERROR_IF(m_incomingPushTransaction, Push, "UI process failed to fetch push before incoming push transaction timed out.");
+    m_incomingPushTransaction = nullptr;
```
Should we remove messages so that we do not jetsam?

bnham replied:
I think probably not, as the process might just be taking a long time to start up. If jetsam really needs to kill us, they should be able to since we are not dirty (no transactions held) and we are in a pretty low jetsam band already.

youennf reviewed (commented):
LGTM overall, some nits

bnham committed:
webpushd should protect incoming pushes with an os_transaction
https://bugs.webkit.org/show_bug.cgi?id=241740

Reviewed by Youenn Fablet.

webpushd needs to acquire an os_transaction in its incoming push message path, or else it might get
jetsamed in the middle of processing a push.

* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::didReceivePushMessage): Create a transaction around fetching the decryption
key from the DB and decrypting the push.

* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::Daemon):
(WebPushD::Daemon::startIncomingPushTransactionTimer):
(WebPushD::Daemon::cancelIncomingPushTransactionTimer):
(WebPushD::Daemon::incomingPushTransactionTimerFired):
(WebPushD::Daemon::handleIncomingPush):
(WebPushD::Daemon::getPendingPushMessages): Create a transaction around waking up the UIProcess to
handle the decrypted push and time it out after 10 seconds.

* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain): Create a transaction around the time from when we spawn to when we
start initializing PushService. (PushService initialization involves async steps and takes care of
acquiring its own transaction in its constructor.)

Canonical link: https://commits.webkit.org/251879@main
